### PR TITLE
Created extendable ES Documents

### LINKED ISSUE #20: Change traits to extendable documents
According to new feature in ElasticSeach bundle https://github.com/ongr-io/ElasticsearchBundle/blob/master/Resources/doc/inheritance.md

### PR DESCRIPTION
**Pull request by MindaugasR**
Created extendable ES Documents;
Document/Traits marked as deprecated;
Updated CategoryService, TwigExtensions and tests;

closes #20

### CODE REVIEW

**Comment by MindaugasR** [comment]
blocked by https://github.com/ongr-io/ElasticsearchBundle/pull/117

**Comment by saimaz** [comment]
What about `Abstract` prefix in abstract classes?

File: Tests/Unit/Document/AbstractUrlObjectTest.php [hunk 1]
```diff
@@ -44,16 +44,16 @@ public function urlTraitDataProvider()
...
-     * @dataProvider urlTraitDataProvider()
+     * @dataProvider urlObjectDataProvider()
      */
-    public function testUrlTrait(array $data)
+    public function testUrlObject(array $data)
     {
-        /** @var UrlObjectTrait $url */
-        $url = $this->getMockForTrait('ONGR\\ContentBundle\\Document\\Traits\\UrlObjectTrait');
+        /** @var AbstractUrlObject $url */
+        $url = $this->getMock('ONGR\ContentBundle\Document\AbstractUrlObject', null);
```
Comment (saimaz) [inline Tests/Unit/Document/AbstractUrlObjectTest.php]: Abstract class should be mocked with `getMockForAbstractClass`

File: Tests/Unit/Document/AbstractCdnObjectTest.php [hunk 2]
```diff
@@ -42,16 +42,16 @@ public function cdnTraitDataProvider()
...
-     * @dataProvider cdnTraitDataProvider()
+     * @dataProvider cdnObjectDataProvider()
      */
-    public function testCdnTrait(array $data)
+    public function testCdnObject(array $data)
     {
-        /** @var CdnObjectTrait $cdn */
-        $cdn = $this->getMockForTrait('ONGR\\ContentBundle\\Document\\Traits\\CdnObjectTrait');
+        /** @var AbstractCdnObject $cdn */
+        $cdn = $this->getMock('ONGR\ContentBundle\Document\AbstractCdnObject', null);
```
Comment (saimaz) [inline Tests/Unit/Document/AbstractCdnObjectTest.php]: Abstract class should be mocked with `getMockForAbstractClass`

File: Tests/Unit/Document/AbstractImageNestedTest.php [hunk 3]
```diff
@@ -46,16 +46,16 @@ public function imagesNestedTraitDataProvider()
...
-     * @dataProvider imagesNestedTraitDataProvider()
+     * @dataProvider imageNestedDataProvider()
      */
-    public function testImagesNestedTrait(array $data)
+    public function testImageNested(array $data)
     {
-        /** @var ImagesNestedTrait $imagesNested */
-        $imagesNested = $this->getMockForTrait('ONGR\\ContentBundle\\Document\\Traits\\ImagesNestedTrait');
+        /** @var AbstractImageNested $imagesNested */
+        $imagesNested = $this->getMock('ONGR\ContentBundle\Document\AbstractImageNested', null);
```
Comment (saimaz) [inline Tests/Unit/Document/AbstractImageNestedTest.php]: Abstract class should be mocked with `getMockForAbstractClass`

**Comment by saimaz** [comment]
@MindaugasR I'm still thinking about nested and object types here. Is this the very best way to make it abstract? Why don't we just include in abstract document as normal class? Most likely that url or image object's will not gonna change and if it will child document class could easily override parameter and set new object class. What do you think?

@tautrimas do you have any concerns about it?

**Comment by MindaugasR** [comment]
@saimaz: I agree that nested and object types it is not necessary to have as abstract classes. And your notes about extending and overriding it when it is necessary was very correct. I will change it.

**Pull request merged**